### PR TITLE
DAOS-3751 tests: Verify snapshot aggregation

### PR DESCRIPTION
**Pull request by phender**
Verify snapshot aggregation with 2 servers and 6 clients (CI limit).
Write the same data to the pool twice.  Create a snapshot between the
writes and confirm that deleting the snapshot reduces the pool capacity
by half.

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

### CODE REVIEW

**Comment by johannlombardi** [comment]
@phender What is the state of affairs on this PR? Is it ready for review?

**Comment by phender** [comment]
> @phender What is the state of affairs on this PR? Is it ready for review?

Waiting for https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-2906/21/pipeline to complete.  It could be reviewed in the meantime I suppose.

File: src/tests/ftest/container/snapshot_aggregation.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            "Pool free space after second write:\n  SCM:  %s\n  NVMe: %s",
+            self.free_space["scm"][-1], self.free_space["nvme"][-1])
+        self.assertLess(
+            self.free_space["scm"][2]["api"],
+            self.free_space["scm"][1]["api"],
+            "SCM free pool space was not reduced by the overwrite")
+        self.assertLess(
+            self.free_space["nvme"][2]["api"],
+            self.free_space["nvme"][1]["api"],
+            "SCM free pool space was not reduced by the overwrite")
```
Comment (shimizukko) [inline src/tests/ftest/container/snapshot_aggregation.py:137]:
```suggestion
            "NVMe free pool space was not reduced by the overwrite")
```

File: src/tests/ftest/container/snapshot_aggregation.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            "Pool free space after first write:\n  SCM:  %s\n  NVMe: %s",
+            self.free_space["scm"][-1], self.free_space["nvme"][-1])
+        self.assertLess(
+            self.free_space["scm"][1]["api"],
+            self.free_space["scm"][0]["api"],
+            "SCM free pool space was not reduced by the initial write")
+        self.assertLess(
+            self.free_space["nvme"][1]["api"],
+            self.free_space["nvme"][0]["api"],
+            "SCM free pool space was not reduced by the initial write")
```
Comment (shimizukko) [inline src/tests/ftest/container/snapshot_aggregation.py:108]:
```suggestion
            "NVMe free pool space was not reduced by the initial write")
```

File: src/tests/ftest/container/snapshot_aggregation.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            "Pool free space after deleting snapshot:\n  SCM:  %s\n  NVMe: %s",
+            self.free_space["scm"][-1], self.free_space["nvme"][-1])
+        self.assertEqual(
+            self.free_space["scm"][3]["api"],
+            self.free_space["scm"][1]["api"],
+            "SCM free pool space was not restored by the aggregation")
+        self.assertEqual(
+            self.free_space["nvme"][3]["api"],
+            self.free_space["nvme"][1]["api"],
+            "SCM free pool space was not restored by the aggregation")
```
Comment (shimizukko) [inline src/tests/ftest/container/snapshot_aggregation.py:159]:
```suggestion
            "NVMe free pool space was not restored by the aggregation")
```

File: src/tests/ftest/container/snapshot_aggregation.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    def update_free_space(self):
+        """Append the free space list with the current pool capacities."""
+        data = self.dmg.pool_query(self.pool.uuid)
+        self.pool.get_info()
+
+        for index, name in enumerate(("scm", "nvme")):
+            self.free_space[name].append({
+                "dmg": data[name]["free"],
+                "api": int(self.pool.info.pi_space.ps_space.s_free[index])
```
Comment (shimizukko) [inline src/tests/ftest/container/snapshot_aggregation.py:54]: We have dmg and api keys, but you use api everywhere. Is there a reason? Are they interchangeable?

**Review by shimizukko — APPROVED** [review]
LGTM

File: src/tests/ftest/container/snapshot_aggregation.py (hunk 4 above)
Comment (phender) [reply src/tests/ftest/container/snapshot_aggregation.py:54]: Originally I was planning on updating the `TestPool.get_info()` to support either the `dmg` or `api` control method and return the same data structure regardless of which control method was used, but the output from the dmg command did not map well and it became a larger-than-the-original-PR issue to resolve for all the tests that make use of the `TestPool.get_info()` method.  So for now I'm calling both methods, using the dmg output to display the free space (human readable) and the API output to do the comparison (values in the same units).

File: src/tests/ftest/container/snapshot_aggregation.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            })
+
+    def detect_aggregation_complete(self):
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/container/snapshot_aggregation.py:57]: Future plans include replacing the steps inside this method with a more deterministic approach to detecting the completion of aggregation - at which point the name will be more appropriate.

File: src/tests/ftest/container/snapshot_aggregation.py (hunk 1 above)
Comment (phender) [reply src/tests/ftest/container/snapshot_aggregation.py:137]: Thanks, good catch.

File: src/tests/ftest/container/snapshot_aggregation.py (hunk 2 above)
Comment (phender) [reply src/tests/ftest/container/snapshot_aggregation.py:108]: Thanks, good catch.

File: src/tests/ftest/container/snapshot_aggregation.py (hunk 3 above)
Comment (phender) [reply src/tests/ftest/container/snapshot_aggregation.py:159]: Thanks, good catch.

File: src/tests/ftest/container/snapshot_aggregation.py (hunk 4 above)
Comment (shimizukko) [reply src/tests/ftest/container/snapshot_aggregation.py:54]: I see. dmg pool query returns a lot of stuff, so I can imagine it's hard to map the output with the API. Thanks.

File: src/tests/ftest/container/snapshot_aggregation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shimizukko) [reply src/tests/ftest/container/snapshot_aggregation.py:57]: I see. Thanks

**Comment by johannlombardi** [comment]
@phender The specific test added by this patch seems to have failed. Could you please have a look?

**Review by shimizukko — APPROVED** [review]
LGTM

**Comment by phender** [comment]
> @phender The specific test added by this patch seems to have failed. Could you please have a loo

**Comment by phender** [comment]
> @phender The specific test added by this patch seems to have failed. Could you please have a look?

We talked about this earlier this morning.  The second ior command failed:
```
2020-09-26 07:33:32,841 output           L0655 DEBUG| Command environment vars:
2020-09-26 07:33:32,842 output           L0655 DEBUG|   {'FI_PSM2_DISCONNECT': '1', 'DAOS_POOL': '32FF5CB6-B1F8-4210-B7D7-7F5ED9EACEA4', 'D_LOG_FILE': '/var/tmp/daos_testing/snapshot_aggregation.py-1_daos_client.log', 'IOR_HINT__MPI__romio_daos_obj_class': 'SX', 'DAOS_SVCL': '0', 'DAOS_CONT': '70F35044-AE31-4AD5-B449-0DD61DD9DFE6', 'MPI_LIB': '""'}
2020-09-26 07:33:32,843 process          L0389 INFO | Running '/usr/lib64/mpich/bin/mpirun -hostfile /var/tmp/avocado_w4hIFo/1-._container_snapshot_aggregation.py_SnapshotAggregation.test_snapshot_aggregation_container-hosts-ior-pool-servers-f0d0/hostfile73938 -np 6 ior -a DFS -b 10G -w -E -k -i 1 -G 456 -o /test_file -t 1M --dfs.chunk_size 1048576 --dfs.cont 70F35044-AE31-4AD5-B449-0DD61DD9DFE6 --dfs.dir_oclass SX --dfs.group daos_server --dfs.oclass SX --dfs.pool 32FF5CB6-B1F8-4210-B7D7-7F5ED9EACEA4 --dfs.svcl 0'
2020-09-26 07:33:33,195 process          L0479 DEBUG| [stdout] [1,3]<stdout>:DEBUG 07:33:33.195217 infocache.go:100: Retrieved response for NUMA 0 with device index 0
2020-09-26 07:33:33,196 process          L0479 DEBUG| [stdout] [1,1]<stdout>:DEBUG 07:33:33.195010 infocache.go:100: Retrieved response for NUMA 0 with device index 0
2020-09-26 07:33:33,196 process          L0479 DEBUG| [stdout] [1,6]<stdout>:DEBUG 07:33:33.195228 infocache.go:100: Retrieved response for NUMA 0 with device index 0
2020-09-26 07:33:33,196 process          L0479 DEBUG| [stdout] [1,5]<stdout>:DEBUG 07:33:33.195231 infocache.go:100: Retrieved response for NUMA 0 with device index 0
2020-09-26 07:33:33,196 process          L0479 DEBUG| [stdout] [1,4]<stdout>:DEBUG 07:33:33.195311 infocache.go:100: Retrieved response for NUMA 0 with device index 0
2020-09-26 07:33:33,196 process          L0479 DEBUG| [stdout] [1,2]<stdout>:DEBUG 07:33:33.195058 infocache.go:100: Retrieved response for NUMA 0 with device index 0
2020-09-26 07:33:33,204 process          L0479 DEBUG| [stdout] [1,1]<stdout>:DEBUG 07:33:33.204307 infocache.go:100: Retrieved response for NUMA 0 with device index 0
2020-09-26 07:33:33,211 process          L0479 DEBUG| [stdout] [1,1]<stdout>:DEBUG 07:33:33.210658 domain_info.go:80: client pid: 25948 uid: 1101 gid 1101 ctx: 
2020-09-26 07:33:33,234 process          L0479 DEBUG| [stdout] IOR-3.3.0+dev: MPI Coordinated Test of Parallel I/O
2020-09-26 07:33:33,234 process          L0479 DEBUG| [stdout] Began               : Sat Sep 26 07:33:33 2020
2020-09-26 07:33:33,234 process          L0479 DEBUG| [stdout] Command line        : ior -a DFS -b 10G -w -E -k -i 1 -G 456 -o /test_file -t 1M --dfs.chunk_size 1048576 --dfs.cont 70F35044-AE31-4AD5-B449-0DD61DD9DFE6 --dfs.dir_oclass SX --dfs.group daos_server --dfs.oclass SX --dfs.pool 32FF5CB6-B1F8-4210-B7D7-7F5ED9EACEA4 --dfs.svcl 0
2020-09-26 07:33:33,235 process          L0479 DEBUG| [stdout] Machine             : Linux wolf-144.wolf.hpdd.intel.com
2020-09-26 07:33:33,235 process          L0479 DEBUG| [stdout] TestID              : 0
2020-09-26 07:33:33,235 process          L0479 DEBUG| [stdout] StartTime           : Sat Sep 26 07:33:33 2020
2020-09-26 07:33:33,241 process          L0479 DEBUG| [stdout] 
2020-09-26 07:33:33,241 process          L0479 DEBUG| [stdout] Options: 
2020-09-26 07:33:33,241 process          L0479 DEBUG| [stdout] api                 : DFS
2020-09-26 07:33:33,241 process          L0479 DEBUG| [stdout] apiVersion          : DAOS
2020-09-26 07:33:33,241 process          L0479 DEBUG| [stdout] test filename       : /test_file
2020-09-26 07:33:33,241 process          L0479 DEBUG| [stdout] access              : single-shared-file
2020-09-26 07:33:33,242 process          L0479 DEBUG| [stdout] type                : independent
2020-09-26 07:33:33,242 process          L0479 DEBUG| [stdout] segments            : 1
2020-09-26 07:33:33,242 process          L0479 DEBUG| [stdout] ordering in a file  : sequential
2020-09-26 07:33:33,242 process          L0479 DEBUG| [stdout] ordering inter file : no tasks offsets
2020-09-26 07:33:33,242 process          L0479 DEBUG| [stdout] nodes               : 6
2020-09-26 07:33:33,242 process          L0479 DEBUG| [stderr] ERROR (aiori-DFS.c:541): 0: 17: dfs_open() of test_file Failed
2020-09-26 07:33:33,243 process          L0479 DEBUG| [stdout] tasks               : 6
2020-09-26 07:33:33,243 process          L0479 DEBUG| [stdout] clients per node    : 1
2020-09-26 07:33:33,243 process          L0479 DEBUG| [stdout] repetitions         : 1
2020-09-26 07:33:33,243 process          L0479 DEBUG| [stdout] xfersize            : 1 MiB
2020-09-26 07:33:33,243 process          L0479 DEBUG| [stdout] blocksize           : 10 GiB
2020-09-26 07:33:33,244 process          L0479 DEBUG| [stdout] aggregate filesize  : 60 GiB
2020-09-26 07:33:33,244 process          L0479 DEBUG| [stdout] 
2020-09-26 07:33:33,244 process          L0479 DEBUG| [stdout] Results: 
2020-09-26 07:33:33,244 process          L0479 DEBUG| [stdout] 
2020-09-26 07:33:33,244 process          L0479 DEBUG| [stdout] access    bw(MiB/s)  IOPS       Latency(s)  block(KiB) xfer(KiB)  open(s)    wr/rd(s)   close(s)   total(s)   iter
2020-09-26 07:33:33,244 process          L0479 DEBUG| [stdout] ------    ---------  ----       ----------  ---------- ---------  --------   --------   --------   --------   ----
2020-09-26 07:33:33,245 process          L0479 DEBUG| [stderr] Fatal error in PMPI_Barrier: Unknown error class, error stack:
2020-09-26 07:33:33,245 process          L0479 DEBUG| [stderr] PMPI_Barrier(253)...................: MPI_Barrier(comm=MPI_COMM_WORLD) failed
2020-09-26 07:33:33,245 process          L0479 DEBUG| [stderr] PMPI_Barrier(239)...................: 
2020-09-26 07:33:33,245 process          L0479 DEBUG| [stderr] MPIR_Barrier_impl(146)..............: 
2020-09-26 07:33:33,246 process          L0479 DEBUG| [stderr] MPIR_Barrier_intra_dissemination(53): Failure during collective
2020-09-26 07:33:33,246 process          L0479 DEBUG| [stderr] Fatal error in PMPI_Barrier: Unknown error class, error stack:
2020-09-26 07:33:33,246 process          L0479 DEBUG| [stderr] PMPI_Barrier(253)..........................: MPI_Barrier(comm=MPI_COMM_WORLD) failed
2020-09-26 07:33:33,246 process          L0499 INFO | Command '/usr/lib64/mpich/bin/mpirun -hostfile /var/tmp/avocado_w4hIFo/1-._container_snapshot_aggregation.py_SnapshotAggregation.test_snapshot_aggregation_container-hosts-ior-pool-servers-f0d0/hostfile73938 -np 6 ior -a DFS -b 10G -w -E -k -i 1 -G 456 -o /test_file -t 1M --dfs.chunk_size 1048576 --dfs.cont 70F35044-AE31-4AD5-B449-0DD61DD9DFE6 --dfs.dir_oclass SX --dfs.group daos_server --dfs.oclass SX --dfs.pool 32FF5CB6-B1F8-4210-B7D7-7F5ED9EACEA4 --dfs.svcl 0' finished with 101 after 0.399180173874s
2020-09-26 07:33:33,247 process          L0479 DEBUG| [stderr] PMPI_Barrier(239)..........................: 
2020-09-26 07:33:33,247 process          L0479 DEBUG| [stderr] MPIR_Barrier_impl(146).....................: 
2020-09-26 07:33:33,247 process          L0479 DEBUG| [stderr] MPIR_Barrier_intra_dissemination(43).......: 
2020-09-26 07:33:33,247 process          L0479 DEBUG| [stderr] MPIDI_CH3U_Complete_posted_with_error(1100): Process failed
2020-09-26 07:33:33,247 process          L0479 DEBUG| [stderr] MPIR_Barrier_intra_dissemination(43).......: 
2020-09-26 07:33:33,247 process          L0479 DEBUG| [stderr] MPIDI_CH3_iSend(24)........................: Communication error with rank 0
2020-09-26 07:33:33,248 process          L0479 DEBUG| [stderr] Fatal error in PMPI_Barrier: Unknown error class, error stack:
2020-09-26 07:33:33,248 process          L0479 DEBUG| [stderr] PMPI_Barrier(253)..........................: MPI_Barrier(comm=MPI_COMM_WORLD) failed
2020-09-26 07:33:33,248 process          L0479 DEBUG| [stderr] PMPI_Barrier(239)..........................: 
2020-09-26 07:33:33,248 process          L0479 DEBUG| [stderr] MPIR_Barrier_impl(146).....................: 
2020-09-26 07:33:33,248 process          L0479 DEBUG| [stderr] MPIR_Barrier_intra_dissemination(43).......: 
2020-09-26 07:33:33,248 process          L0479 DEBUG| [stderr] MPIDI_CH3U_Complete_posted_with_error(1100): Process failed
2020-09-26 07:33:33,248 process          L0479 DEBUG| [stderr] Fatal error in PMPI_Barrier: Unknown error class, error stack:
2020-09-26 07:33:33,249 process          L0479 DEBUG| [stderr] PMPI_Barrier(253)..........................: MPI_Barrier(comm=MPI_COMM_WORLD) failed
2020-09-26 07:33:33,249 process          L0479 DEBUG| [stderr] PMPI_Barrier(239)..........................: 
2020-09-26 07:33:33,249 process          L0479 DEBUG| [stderr] MPIR_Barrier_impl(146).....................: 
2020-09-26 07:33:33,249 process          L0479 DEBUG| [stderr] MPIR_Barrier_intra_dissemination(43).......: 
2020-09-26 07:33:33,249 process          L0479 DEBUG| [stderr] MPIDI_CH3U_Complete_posted_with_error(1100): Process failed
2020-09-26 07:33:33,249 process          L0479 DEBUG| [stderr] Fatal error in PMPI_Barrier: Unknown error class, error stack:
2020-09-26 07:33:33,249 process          L0479 DEBUG| [stderr] PMPI_Barrier(253)...................: MPI_Barrier(comm=MPI_COMM_WORLD) failed
2020-09-26 07:33:33,250 process          L0479 DEBUG| [stderr] PMPI_Barrier(239)...................: 
2020-09-26 07:33:33,250 process          L0479 DEBUG| [stderr] MPIR_Barrier_impl(146)..............: 
2020-09-26 07:33:33,250 process          L0479 DEBUG| [stderr] MPIR_Barrier_intra_dissemination(53): Failure during collective
```
I asked Mohamad, as you had suggested, and said it was a different issue and he haven't seen a failed barrier with this stack in mpich.

Due to another issue (avocado tags including "-") the run did not include results for other ior tests to compare.  I'm rerunning the PR with the avocado test tag fix to see if the issue was intermittent.

**Review by shimizukko — APPROVED** [review]
LGTM

**Review by shimizukko — APPROVED** [review]
LGTM

File: src/tests/ftest/container/snapshot_aggregation.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        """Detect that aggregation has started and completed."""
+        # Detect aggregation start
+        self.log.info("Waiting for 90 seconds for aggregation to start")
+        time.sleep(90)
+
+        # Detect aggregation completion
+        self.log.info("Waiting for 30 seconds for aggregation to finish")
+        time.sleep(30)
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_aggregation.py:62]: I would recommend looping in smaller chunks rather just going to sleep in one go for 120 secs.
Something like 60 secs of loop and check back and if not complete do it again. You may decide how many times you would like to loop. 2-3 seems reasonable to me.
The aggregation completion time here is somewhat hardcoded. And this value can change in future due to either performance improvement or degradation. I believe if the aggregation finishes in 60 secs we will be wasting 1 minute on all runs for no reason. But if it doesn't finish in 60 secs we can loop back and check it again and if still doesn't finish we can fail the test.

**Review by saurabhtandan — CHANGES_REQUESTED** [review]

File: src/tests/ftest/container/snapshot_aggregation.py (hunk 6 above)
Comment (phender) [reply src/tests/ftest/container/snapshot_aggregation.py]: Done.

File: src/tests/ftest/container/snapshot_aggregation.py [hunk 7]
```diff
@@ -134,26 +118,48 @@ def test_snapshot_aggregation(self):
...
+            # Update the utilized capacity of the pool
+            self.pool.get_info()
+            self.pool.set_query_data()
+            self.update_free_space()
+            self.log.info(
+                "Pool free space %s seconds after deleting the snapshot:"
+                "\n  SCM:  %s\n  NVMe: %s", sleep_time * loop_count,
+                self.free_space["scm"][-1], self.free_space["nvme"][-1])
+
+            # Determine if the utilized NVMEe capacity of the pool has been
```
Comment (shimizukko) [inline src/tests/ftest/container/snapshot_aggregation.py:150]:
```suggestion
            # Determine if the utilized NVMe capacity of the pool has been
```

**Review by shimizukko — APPROVED** [review]
LGTM

File: src/tests/ftest/container/snapshot_aggregation.py (hunk 7 above)
Comment (phender) [reply src/tests/ftest/container/snapshot_aggregation.py:150]: Thanks

Comment (phender) [reply src/tests/ftest/container/snapshot_aggregation.py]: Done.

**Review by shimizukko — APPROVED** [review]
LGTM

**Comment by phender** [comment]
Failures unrelated to changes in this PR in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-2906/36/:
- daos_test/daos_core_test (not marked): https://jira.hpdd.intel.com/browse/DAOS-5783
- daos_test-daos_core_test-rebuild: https://jira.hpdd.intel.com/browse/DAOS-5784
- checksum/csum_error_logging.py: https://jira.hpdd.intel.com/browse/DAOS-5745
- io/fio-small.py: https://jira.hpdd.intel.com/browse/DAOS-5780

**Review by saurabhtandan — APPROVED** [review]
LGTM

**Pull request merged**